linera-ethereum/src/client.rs

```diff
@@ -1,28 +1,27 @@
 // Copyright (c) Zefchain Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use std::time::Duration;
-
-use ethers::{
-    core::types::Bytes,
-    prelude::{Http, Provider},
-    providers::JsonRpcClient,
-    types::{NameOrAddress, TransactionRequest, U256},
+use alloy::{
+    primitives::{Address, U256},
+    providers::{Provider, ProviderBuilder, RootProvider},
+    rpc::types::eth::{
+        request::{TransactionInput, TransactionRequest},
+        BlockId, BlockNumberOrTag, Filter,
+    },
 };
-use ethers_core::types::{Address, BlockId, BlockNumber, Filter};
-use ethers_middleware::Middleware;
+use alloy_primitives::Bytes;
+use reqwest::Client;
 
 use crate::common::{event_name_from_expanded, parse_log, EthereumEvent, EthereumServiceError};
+pub type HttpProvider = RootProvider<alloy::transports::http::Http<Client>>;
 
 /// The Ethereum endpoint and its provider used for accessing the ethereum node.
 pub struct EthereumEndpoint<M> {
-    pub provider: Provider<M>,
+    pub provider: M,
 }
 
-impl<M> EthereumEndpoint<M>
-where
-    M: JsonRpcClient,
-{
+//
+impl EthereumEndpoint<HttpProvider> {
```

Comment:
Should we remove the type parameter entirely then? Did you mean to use a trait?

Reply:
OK so I just read this comment in your summary:
> For some unclear reason, the impl<M> EthereumEndpoint with M: Provider does not work 

Seems like you need to add a `// TODO(#NUM)` then, so that we investigate later.

Reply:
Ok, done.